packages/mobx-state-tree/src/types/complex-types/model.ts

```diff
@@ -138,18 +150,49 @@ export class ModelType<S, T> extends ComplexType<S, T> implements IModelType<S,
         // ensures that any default value gets converted to its related type
         this.properties = toPropertiesObject(this.properties)
         freeze(this.properties) // make sure nobody messes with it
-    }
+        this.propertyNames = Object.keys(this.properties)
+        this._preBootModel()
+    }
+
+    private _preBootModel() {
+        let optionalFound = false
```

Comment:
would it matter to just store all properties in the optionalMap? (and by that not having optional map at all but just look at the properties if something from the snapshot?)

Reply:
If I ubderstand correctly, you suggest to replace 
```typescript
        if (processed && this.optionalChildren) {
            const optionalChildren = this.optionalChildren
            Object.keys(optionalChildren).forEach(name => {
                if (typeof processed[name] === "undefined") {
                    processed[name] = optionalChildren[name].getDefaultValueSnapshot()
                }
            })
        }
```
to
```typescript
        if (processed) {
            this.forAllProps((name, type) => {
                if (typeof processed[name] === "undefined") {
                    const optional = tryGetOptional(type)
                    if (optional) {
                        processed[name] = optional.getDefaultValueSnapshot()
                    }
                }
            })
        }
``` 
at ```applySnapshotPreProcessor```. By doing so we can avoid introducing ```this.optionalChildren
```.

It will be a little bit more cpu-intensive, as ```_preBootModel``` happens once for type and ```applySnapshotPreProcessor``` happens on every snapshot application. 
But on the other hand code will be clearer and couple of ```typeof``` checks do not seem heavy. Revisiting this code also made me think that ```types.late(() => types.optional(...))) case is not covered. It's quite strange and rare, but possible.
I will add changes within couple of hours ;)